[RFC] Migrate legacy test 74.

lucc opened the pull request:
Currently the test fails.  I guess that the problem is related to the fact that the original test redirects messages to test.out.  I have not yet found a suitable way of doing this (it seems).

As this is the first test I migrate I am happy about any help.

lucc committed:
tests: Migrate legacy test 74.

Helped-By: David Bürgin <676c7473@gmail.com>
Helped-By: Justin M. Keyes <justinkz@gmail.com>

lucc commented:
I am having trouble to get this test to work. Can anybody tell me how I can get some more error messages out of the test run or in which directory the tests are run?

justinmk commented:
>  I guess that the problem is related to the fact that the original test redirects messages to test.out.

`redir >> test.out` should work fine, if that's what you mean.

If the test is hanging, it may be waiting for user input. Remove lines until it doesn't hang, then figure out how to avoid the prompt. 

To run just this individual test:

    TEST_FILE=test/functional/legacy/074_global_var_in_viminfo_spec.lua make functionaltest



Also just a reminder to check https://github.com/neovim/neovim/wiki/Unit-tests when you get a chance.

lucc commented:
Thanks for the link. I also found #2146 helpful in order to only run one test. (`TEST_FILTER=74 make functionaltest` works for me)

The test should create a file `Xviminfo` but I don't know the working directory of the test. How can I see the working dir of the test (or the file for that matter)?

justinmk commented:
lfs.currentdir()

lucc committed:
tests: Add tests for the :wv command.

lucc commented:
Thanks, with `lfs.currdir()` I found that the test is run in the same directory as `make`. I already suspected this when files which I `:w`rote in tests were there after the test run.

*But* the viminfo file is still not there, so I added a new test file (`test/functional/legacy/write_viminfo_spec.lua`) which illustrates the error. This new test also fails but I am pretty sure it should not. Does anyone see why it fails, or rather why the viminfo file is not written?

glts commented:
Probably because the tests are run with `-i NONE`, see `:h -i`.

lucc commented:
Thanks @glts that would explain it. I just tried to dig my way though the makefile but to no avail. Do you know where the command line for the tests is defined? 

Then I can at least test if this will work if we don't use `nvim -i NONE`. The next step could be to discuss if and how to change it. Or what to do with this test otherwise.

glts commented:
It's in `test/functional/helpers.lua`. My guess is that you would have to create a custom session to run this test in.

If you can't get it to work there are easier candidates for conversion :) I was planning to convert some more tests myself perhaps I can take a look at this test at some point later.

lucc commented:
Ok I had a look at the helpers.lua file but did not manage to see how I could create a custom session. The file returns a `set_session` function which *I think* I should use but I have not seen anything there how sessions are *created*.

Is there some documentation for helpers.lua, where you can point me to? Else I will give up for the moment.

justinmk commented:
> did not manage to see how I could create a custom session

grepping the other tests usually turns up a couple good examples.

    local nvim2 = helpers.spawn({helpers.nvim_prog, '-u', 'NONE', '--embed'})
    helpers.set_session(nvim2)


> The file returns a set_session function which I think I should use but I have not seen anything there how sessions are created.

Yes, that's a hack I added for `recover_spec.lua`, to test *two* sessions in one test. 

> Is there some documentation for helpers.lua, where you can point me to?

There isn't yet. The non-legacy tests are pretty easy to learn from as examples, though.

splinterofchaos commented on test/functional/legacy/074_global_var_in_viminfo_spec.lua:
```diff
@@ -0,0 +1,70 @@
+    execute('rv! Xviminfo')
+    -- TODO: Actually we want to test if the viminfo file was written but the
+    -- test is currently run with "nvim -i NONE ..." which inhibits this.  So
+    -- the test would fail here.
```
At the start of this test, you can put
```lua
os.remove("Xviminfo")  -- ensure it does not exist
```
and between the `wv` and `rv` commands, put
```lua
neq(nil, lfs.attributes('Xviminfo'))  -- assert it exists
```
Instead of using `redir > test.out` below (since that's not what's being tested), you can just put
```lua
eq({'foo'=1, 'longvariable'=1000, 'bar'=0}, eval('MY_GLOBAL_DICT'))
```
I could have the syntax a little wrong, but that's the general idea.

lucc commented:
@justinmk thanks for the hint about set_session. I'm still having trouble with it (sadly grepping did only turn up one example).

I added the file `test/functional/ex_cmds/wviminfo_spec.lua` in the last commit. But the first test therein fails or passes randomly (it seems to fail more often than it passes for me). I do not understand why. (The second test therein currently fails.) Sometimes the file `foo` still exists after the test run although it should be deleted before and after each test and at teardown.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,54 @@
+    neq(timestamp, lfs.attributes(file, 'modification'))
+  end)
+
+  teardown(function() os.remove(file) end)
```
This is redundant with the `after_each` block. Note that `teardown` and `setup` run before/after each `describe` group, *not* after *each test* (`it` blocks). This is kind of confusing for people used to e.g. junit. See http://olivinelabs.com/busted/ for specifics.

`before_each` and `after_each` will run before/after each `it` test.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,54 @@
+  helpers.nvim_prog, helpers.set_session
+
+describe(':wviminfo', function()
+  local file = './foo'
```
might want to use `lfs.currentdir()` just to be explicit.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,54 @@
+    clear()
+    os.remove(file)
+  end)
+  after_each(function() os.remove(file) end)
```
This is redundant with `before_each` since it's doing the same thing, and `before_each` will run before each `it` test.

lucc commented:
Fixed and rebased. The tests now pass for me. Two points should be noted:

1. I added a test file in `test/functional/ex_commands` to test `:wv`. There is one TODO item in this file and there are may more things that could be tested for `:wv`. This file was created to debug some failures. It can but need not be included.
2. While I was figuring out the various cases for failure I incrementally removed the old way of testing form the converted test 74 (that is adding text to the current buffer and `expect`ing it). Instead the test now uses the `eq`, `neq` and `eval` functions. Some more precise testing (or rather more precise failures) is possible with that but the test is not as close to the original test74 as some other migrated legacy tests. If that is not desired I can try to change that.

justinmk commented:
Restarted build (was a gcov quirk).

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,62 @@
+  local round = math.floor(sec)
+  if round == 0 then round = 1 end
+  os.execute('sleep '..sec..' || sleep '..round)
+end
```
I don't see where this sleep function is actually used, but for future reference: would need to wrap this in a `pending()` condition if the environment (e.g. Windows) doesn't have a `sleep` command. Or better yet, use the Vim `:sleep` command if possible.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,62 @@
+  it('creates a file', function()
+    -- TODO
+    -- Set up the nvim session to be able to write viminfo files.  Is it
+    -- possible to do this outside of the it() call?
```
Since clear() sets up a new session, doing it outside of the it() call wouldn't work. That's why I can `set_session` a hack--ideally, session.lua should support multiple instances. Currently it does not (it's a minor refactoring to do so, just haven't got round to it).

lucc replied:
I just read `helpers.lua` some more and saw that `clear()` does actually start a new session. So now I am doing the same manually.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,62 @@
+    execute('wv! '..file)
+    wait()
+    -- Assert that the file does exist.
+    neq(nil, lfs.attributes(file))
```
Instead of using `eq`/`neq` with a comment explaining the meaning, I tend to use `assert(<condition>, "explanation")`, for example:

    assert(nil ~= lfs.attributes(file), "viminfo file was not found after :wv")

lucc replied:
I tried it but the "explanation" is never shown in the test output. Not for failed and not for passed tests. So what is the advantage of this? I don't think it is a good idea to use an argument which I was unable to find in [the busted docs](http://olivinelabs.com/busted/).

justinmk replied:
> but the "explanation" is never shown in the test output

Are you sure? That would be a major problem. If I change `~=` to `==` at this line: https://github.com/neovim/neovim/commit/1f0830f7004c60f899fc169baecd68960f26ea70#diff-d587b4ef5d4b7947c66765d46ff9a341R50

The test results are:

```
Failure → test/functional/ex_cmds/wviminfo_spec.lua @ 32
:wviminfo overwrites existing files
test/functional/ex_cmds/wviminfo_spec.lua:50: wviminfo_test should be a viminfo-formatted file 
```

>  I don't think it is a good idea to use an argument which I was unable to find in the busted docs.

It's actually a stock lua function. Personally I find it annoying that busted doesn't provide an "explanation" parameter for all of their assert functions. I assume it has to do with BDD dogma about `it()` blocks being small and self-explanatory :P  But if a comment is needed to explain an assertion, then evidently that spec is not self-explanatory (and in practice this is quite common, and not reasonable to avoid it).

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,62 @@
+end
+
+describe(':wviminfo', function()
+  local file = 'foo'
```
This needs a more helpful name like `'wviminfo_test'`.

justinmk commented on test/functional/legacy/074_global_var_in_viminfo_spec.lua:
```diff
@@ -0,0 +1,59 @@
+  setup(function()
+    clear()
+    os.remove("Xviminfo")
+  end)
```
Although it doesn't matter in this case since there's only 1 `it()`, it's a bit safer to use before_each and after_each in case a test is added later.

justinmk commented on test/functional/legacy/074_global_var_in_viminfo_spec.lua:
```diff
@@ -0,0 +1,59 @@
+    eq(test_dict, eval('MY_GLOBAL_DICT'))
+    eq(test_list, eval('MY_GLOBAL_LIST'))
+    -- Assert that the viminfo file does not exists.
+    eq(nil, lfs.attributes('Xviminfo'))
```
Although sanity checks are good, this one is particularly redundant: both setup() and teardown() already removed the file.

justinmk commented on test/functional/legacy/074_global_var_in_viminfo_spec.lua:
```diff
@@ -0,0 +1,59 @@
+    execute('rv! Xviminfo')
+    -- Assert that the variables where restored.
+    eq(1, eval('exists("MY_GLOBAL_DICT")'))
+    eq(1, eval('exists("MY_GLOBAL_LIST")'))
```
Don't need these sanity checks because the next lines test the actual values.

justinmk commented on test/functional/legacy/074_global_var_in_viminfo_spec.lua:
```diff
@@ -0,0 +1,59 @@
+    -- Assert that the viminfo file exists.
+    neq(nil, lfs.attributes('Xviminfo'))
+    execute('unlet MY_GLOBAL_DICT',
+	    'unlet MY_GLOBAL_LIST')
```
use spaces for indent

justinmk commented:
Merged, thanks! I pushed an extra commit containing my suggestions, didn't want to hold up this PR longer. 

FYI, here's how you can override the test session outside of each `it()` block: https://github.com/neovim/neovim/commit/1f0830f7004c60f899fc169baecd68960f26ea70#diff-d587b4ef5d4b7947c66765d46ff9a341R10

Shougo commented:
:+1:

lucc commented:
Thanks @justinmk. I was just to slow to apply your comments :smile:  I was half way through amending when I saw you merged it.

splinterofchaos commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -0,0 +1,62 @@
+  local sec = millisec / 1000
+  local round = math.floor(sec)
+  if round == 0 then round = 1 end
+  os.execute('sleep '..sec..' || sleep '..round)
```
Why use `||` here? `sleep` can't fail, right? Plus, not all shells (or maybe just, annoyingly, fish) support this syntax.

justinmk replied:
@splinterofchaos Just in case you missed my comment above: https://github.com/neovim/neovim/pull/2585#discussion_r32288971

splinterofchaos replied:
Ah, yes. Thanks.
